pkg/controller/namespace/namespace_controller_utils.go

```diff
@@ -103,99 +130,135 @@ func finalizeNamespaceFunc(kubeClient clientset.Interface, namespace *api.Namesp
 	return namespace, err
 }
 
-// deleteAllContent will delete all content known to the system in a namespace. It returns an estimate
-// of the time remaining before the remaining resources are deleted. If estimate > 0 not all resources
-// are guaranteed to be gone.
-// TODO: this should use discovery to delete arbitrary namespace content
-func deleteAllContent(kubeClient clientset.Interface, versions *unversioned.APIVersions, namespace string, before unversioned.Time) (estimate int64, err error) {
-	err = deleteServiceAccounts(kubeClient, namespace)
-	if err != nil {
-		return estimate, err
-	}
-	err = deleteServices(kubeClient, namespace)
-	if err != nil {
-		return estimate, err
-	}
-	err = deleteReplicationControllers(kubeClient, namespace)
-	if err != nil {
-		return estimate, err
-	}
-	estimate, err = deletePods(kubeClient, namespace, before)
-	if err != nil {
-		return estimate, err
-	}
-	err = deleteSecrets(kubeClient, namespace)
-	if err != nil {
-		return estimate, err
-	}
-	err = deleteConfigMaps(kubeClient, namespace)
-	if err != nil {
-		return estimate, err
-	}
-	err = deletePersistentVolumeClaims(kubeClient, namespace)
-	if err != nil {
-		return estimate, err
-	}
-	err = deleteLimitRanges(kubeClient, namespace)
-	if err != nil {
-		return estimate, err
-	}
-	err = deleteResourceQuotas(kubeClient, namespace)
-	if err != nil {
-		return estimate, err
-	}
-	err = deleteEvents(kubeClient, namespace)
-	if err != nil {
-		return estimate, err
-	}
-	// If experimental mode, delete all experimental resources for the namespace.
-	if containsVersion(versions, "extensions/v1beta1") {
-		resources, err := kubeClient.Discovery().ServerResourcesForGroupVersion("extensions/v1beta1")
+// deleteAllContent will use the dynamic client to delete each resource identified in groupVersionResources.
+// It returns an estimate of the time remaining before the remaing resources are deleted.
+// If estimate > 0, not all resources are guaranteed to be gone.
+func deleteAllContent(
+	kubeClient clientset.Interface,
+	clientPool dynamic.ClientPool,
+	opCache operationNotSupportedCache,
+	groupVersionResources []unversioned.GroupVersionResource,
+	namespace string,
+	namespaceDeletedAt unversioned.Time,
+) (int64, error) {
+	estimate := int64(0)
+	glog.V(4).Infof("namespace controller - deleteAllContent - namespace: %s, resources: %v", namespace, groupVersionResources)
+	// iterate over each group version, and attempt to delete all of its resources
+	for _, groupVersionResource := range groupVersionResources {
+		glog.V(4).Infof("namespace controller - deleteAllContent - deleting resource - namespace: %s, resource: %v", namespace, groupVersionResource)
+		// estimate how long it will take for the resource to be deleted (needed for objects that support graceful delete)
+		groupVersionResourceEstimate, supportsGrace, err := estimateGracefulTermination(kubeClient, groupVersionResource, namespace, namespaceDeletedAt)
```

Comment:
I think this all makes total sense.  I will update tomorrow and poke the PR.

On Monday, February 29, 2016, Daniel Smith <notifications@github.com> wrote:

> In pkg/controller/namespace/namespace_controller_utils.go
> <https://github.com/kubernetes/kubernetes/pull/21400#discussion_r54480441>
> :
>
> > +// If estimate > 0, not all resources are guaranteed to be gone.
> > +func deleteAllContent(
> > +	kubeClient clientset.Interface,
> > +	clientPool dynamic.ClientPool,
> > +	opCache operationNotSupportedCache,
> > +	groupVersionResources []unversioned.GroupVersionResource,
> > +	namespace string,
> > +	namespaceDeletedAt unversioned.Time,
> > +) (int64, error) {
> > +	estimate := int64(0)
> > +	glog.V(4).Infof("namespace controller - deleteAllContent - namespace: %s, resources: %v", namespace, groupVersionResources)
> > +	// iterate over each group version, and attempt to delete all of its resources
> > +	for _, groupVersionResource := range groupVersionResources {
> > +		glog.V(4).Infof("namespace controller - deleteAllContent - deleting resource - namespace: %s, resource: %v", namespace, groupVersionResource)
> > +		// estimate how long it will take for the resource to be deleted (needed for objects that support graceful delete)
> > +		groupVersionResourceEstimate, supportsGrace, err := estimateGracefulTermination(kubeClient, groupVersionResource, namespace, namespaceDeletedAt)
>
> I don't know if I want to block this PR or not, but this function would be
> a lot more understandable (and bug free?) if:
>
>    - Move the body of this for loop to another function
>    - Move every cache-check/operation section to its own function
>    - Add a function to the cache like "isSupported" instead of the
>    !cache[op] pattern.
>
> In particular, I kind of think that many of the returns in this function
> body are actually bugs?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21400/files#r54480441>.
>